client/src/main/java/io/pravega/client/byteStream/impl/BufferedByteStreamWriterImpl.java

```diff
@@ -0,0 +1,93 @@
+/**
+ * Copyright (c) 2018 Dell Inc., or its subsidiaries. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ */
+package io.pravega.client.byteStream.impl;
+
+import com.google.common.annotations.VisibleForTesting;
+import io.pravega.client.byteStream.ByteStreamWriter;
+import java.io.IOException;
+import java.nio.ByteBuffer;
+import java.util.concurrent.atomic.AtomicReference;
+import javax.annotation.concurrent.GuardedBy;
+import lombok.RequiredArgsConstructor;
+
+@RequiredArgsConstructor
+public class BufferedByteStreamWriterImpl extends ByteStreamWriter {
+
+    @VisibleForTesting
+    public static final int BUFFER_SIZE = 4096;
+    private final ByteStreamWriterImpl out;
+
+    @GuardedBy("this") //The ref is not guarded but the buffer inside the ref is.
+    private final AtomicReference<ByteBuffer> buffer = new AtomicReference<>(null);
+
+    @Override
+    public void write(int b) throws IOException {
+        synchronized (this) {
+            ByteBuffer localBuffer = buffer.get();
+            if (localBuffer == null) {
+                localBuffer = ByteBuffer.allocate(BUFFER_SIZE);
+                buffer.set(localBuffer);
+            }
+            localBuffer.put((byte) b);
+            if (!localBuffer.hasRemaining()) {
+                flushBuffer();
+            }
+        }
+    }
+
+    @Override
+    public void write(ByteBuffer src) throws IOException {
+        flushBuffer();
+        out.write(src);
+    }
+
+    @Override
+    public void write(byte[] b, int off, int len) throws IOException {
+        flushBuffer();
+        out.write(b, off, len);
+    }
+
+    private void flushBuffer() throws IOException {
+        if (buffer.get() != null) {
+            synchronized (this) {
+                ByteBuffer toWrite = buffer.getAndSet(null);
```

Comment:
We could. I was trying to avoid the 'sync' and allocating the buffer ever in the event that the caller never writes one byte at a time. Because seriously who does that?

I could just accept that the buffer is always going to be there and use it for not only one byte writes but also small ones. Does that sound like a better approach?